File: cpp/algorithms/math/extended_gcd/extended_gcd.cpp
// 扩展欧几里得：模逆与线性同余
#include <alg_std.hpp>
#include <cassert>
using namespace std;

tuple<long long, long long, long long> extgcd(long long a, long long b) {
    if (!b) {
        long long s = a >= 0 ? 1 : -1;
        return {abs(a), s, 0};
    }
    auto [g, x1, y1] = extgcd(b, a % b);
    return {g, y1, x1 - (a / b) * y1};
}

optional<long long> modInverse(long long a, long long mod) {
    auto [g, x, y] = extgcd(a % mod, mod);
  (void)y;
    if (g != 1) return nullopt;
    return (x % mod + mod) % mod;
}

optional<pair<long long, long long>> solveLinearCongruence(long long a, long long b, long long m) {
    auto [g, x0, y] = extgcd(a % m, m);
  (void)y;
    if (b % g != 0) return nullopt;
    long long modM = m / g;
    x0 = ((__int128)x0 * (b / g)) % modM;
    if (x0 < 0) x0 += modM;
    return make_pair(x0, modM);
}

int main() {
    auto [g, x, y] = extgcd(35LL, 15LL);
    assert(g == 5 && 35 * x + 15 * y == 5);

    assert(modInverse(3, 11) == 4);
    assert(!modInverse(4, 6).has_value());

    auto sol = solveLinearCongruence(3, 2, 7);
    assert(sol.has_value());
    auto [x0, period] = *sol;
    assert((3 * x0 - 2) % 7 == 0);
    cout << "extended_gcd OK" << endl;
    return 0;
}

File: python/algorithms/math/extended_gcd/extended_gcd.py
"""扩展欧几里得：ax+by=gcd(a,b)；可解线性同余与模逆。"""

from __future__ import annotations


def extgcd(a: int, b: int) -> tuple[int, int, int]:
    """返回 (g, x, y) 满足 ax + by = g = gcd(a, b)。"""
    if b == 0:
        return abs(a), 1 if a >= 0 else -1, 0
    g, x1, y1 = extgcd(b, a % b)
    return g, y1, x1 - (a // b) * y1


def mod_inverse(a: int, mod: int) -> int | None:
    """a 与 mod 互素时返回 x 使 ax ≡ 1 (mod mod)；否则 None。"""
    g, x, _ = extgcd(a % mod, mod)
    if g != 1:
        return None
    return x % mod


def solve_linear_congruence(a: int, b: int, m: int) -> tuple[int, int] | None:
    """求最小非负 x 使 ax ≡ b (mod m)；无解返回 None，否则 (x0, period)。"""
    g, x0, _ = extgcd(a % m, m)
    if b % g != 0:
        return None
    mod_m = m // g
    x0 = (x0 * (b // g)) % mod_m
    return x0, mod_m


if __name__ == "__main__":
    g, x, y = extgcd(35, 15)
    assert g == 5 and 35 * x + 15 * y == 5

    assert mod_inverse(3, 11) == 4
    assert mod_inverse(4, 6) is None

    sol = solve_linear_congruence(3, 2, 7)
    assert sol is not None
    x0, period = sol
    assert (3 * x0 - 2) % 7 == 0
    print("extended_gcd OK")
